Fog of war

r-a-cristian-93 opened the pull request:
Hi,

This is my attempt at implementing fog of war in flare-game. The development of this new feature is not complete and i just want some feedback to know if i'm on the right track.  I do not expect this request to be merged yet.

The fog of war is tile based and available for isometric mode only. I intend to add corner rounding and neighbour-aware algorithm in the future, also support for orthogonal mode too.

How it works:

- The avatar has a new property "sight" which represents how far he can see expressed in number of tiles. Currently set by default to 8. Can not be edited in mod files, this will change in the future.
- fogofwar=true/false -> in engine/misc.txt turns on/of the fog of war.
- save_fogofwar=true -> in engine/misc.txt turns on/of the fog of war state for each map (explored tiles) - saves the progress.
- Each map file can have a new layer of type "fogofwar" containing default explored tiles (0 is not required).
0 - in avatar's sight
1 - explored
2 - explored but out of sight
```
[layer]
type=fogofwar
data=
1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
1,1,1,1,1,1,1,1,1,1,1,0,0,0,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,
1,1,1,1,1,1,1,1,1,1,1,0,0,0,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,
1,1,1,1,1,1,1,1,1,1,1,0,0,0,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
```
- If the map file has no "fogofwar" layer defined then all the tiles will be set by default to 1 - unvisited;
- Tile look can be defined like any other tiles in "/tilesetdefs/tileset_fogofwar.txt" and "/images/tilesets/tileset_fogofwar.png"
- The minimap does not render unvisited tiles(walls) and does not render out of sight enemies;
- The FogOfWar class holds the tileset and updates the "fogofwar" layer state.
- The assets were temporarily put in a new mod directory called "fogofwar" but i believe they could be moved in the "default" mod directory.
- I did not experience any performance issues on my computer.

Known problems and other improvements that can be done:
- tall buildings/walls/trees tips can be visible at the edge of the map even if the tiles are not visited.
- fog of war should be aware of walls and should not make visible the tiles if the avatar's sight is obstructed.

![fow](https://user-images.githubusercontent.com/69315110/126552065-5b0a1cfa-53b2-4f32-a582-2eb8d4c17a31.jpg)

r-a-cristian-93 commented on src/GameStatePlay.cpp:
```diff
@@ -327,6 +333,7 @@ void GameStatePlay::checkTeleport() {
 			inpt->lock_all = (teleport_mapname == "maps/spawn.txt");
 			mapr->executeOnMapExitEvents();
 			showLoading();
+			save_load->saveGame();
```
Save game when map changes in order to save fog of war state.

dorkster commented:
Instead of drawing a fog layer, I tried taking a different approach by tinting the existing tiles: [Diff](https://github.com/r-a-cristian-93/flare-engine/compare/fog_of_war...dorkster:fog_of_war)

Pros:

- No need for additional assets. So this also works for orthogonal maps.
- Tall tiles get tinted instead of sticking out from beneath the fog layer

Cons:

- Tiles with a large footprint (larger than 1x1, e.g. flare-game's dungeon stairs) don't have a smooth transition between in-sight and out-of-sight

Thoughts on this approach?

dorkster commented on cam:
```diff
@@ -0,0 +1,279 @@
+src/EventManager.cpp:732:	// When an event is delayed, we create a copy and push it to mapr->delayed_events.
```
I'm guessing this file was accidentally committed?

r-a-cristian-93 replied:
That's right, i forgot to remove it.

r-a-cristian-93 replied:
That's right, i forgot to remove it. It shouldn't be there.

r-a-cristian-93 commented:
I like this idea and the way it simplifies things so for. Tall objects that have a footprint of 1x1 look nice as in reality if you can see the base of the object then you can see it's tip too.

I was planning to add neighbour-aware tiles to be able to round the corners for different situations. Supposedly we find a way to handle objects with a footprint larger than 1x1 would it be possible to round the corners of the fog easily? If not, then what about keeping the TileSet tset member of FogOfWar class and create two functions "Sprite makeIsoRectangle()" and "Sprite makeOrthoRectangle()" and use them as default tiles? For basic fog of war this will not require any assets and allows modders to use their custom tiles. As a different flavor modders could use multiple tiles to be able to round the corners.

I believe it would be kind of difficult to manipulate the sprite in order to tint it partially.

dorkster commented:
If you're planning on doing neighbor-aware tiles, you're probably better off sticking with your original approach. For tall tiles, an alternative could be to skip rendering them if all of the following are met:

a. The tile is not in player sight
b. The tile's height (obtained from `MapRenderer::getTileBounds()`) is greater than `eset->tileset.tile_h`

My method could still be the default if there is not fog-of-war tileset defined. It also has better performance (drawing a single tile vs blending two tiles), so it may be worth having a video option to force this "low-quality" fog.

On the topic of performance, it's worth mentioning that the mini-map's `prerender()` can be expensive. I tried to optimize a bit by having it only run when new tiles are discovered: https://github.com/dorkster/flare-engine/commit/30ebb99f57b1a8670d8227fe03d7d28195665d72

r-a-cristian-93 commented:
Hi,

I didn't know how to merge your changes and be able to keep my previous code too in order have two different types of fog of war. I just copy-pasted the code. If you could point me to a guide on how to do it i will undo the last commit and merge it correctly.

dorkster commented:
>I didn't know how to merge your changes and be able to keep my previous code too in order have two different types of fog of war. I just copy-pasted the code. If you could point me to a guide on how to do it i will undo the last commit and merge it correctly.

I don't think there's a way to cleanly merge the two, so what you did here is fine.

r-a-cristian-93 commented:
I was thinking of a possible solution for the large footprint tiles. Maybe we could split them in one direction when the map loads. So instead of a 2x3 (x, y) tile we will have two 1x3 tiles. Do you think it can be done?

dorkster commented:
In order to be truly effective, I think you would need to break them down to 1x1 tiles. I don't think this is worth doing in-engine, and is better left to the game date side of things.

r-a-cristian-93 commented:
I have added fog of war TYPE_MINIMAP (3b3bfa5dd5bfb2a2c05156193de53d59aaef4ae7) to address this request https://github.com/flareteam/flare-engine/issues/1741

r-a-cristian-93 commented:
You were right the minimap's ` prerender()` function is very expensive, it made made the game lag on my system. But i have found a way to improve the minimap performance.

In MenuMiniMap class I have created `update()`, `updateIso() `and` updateOrtho()` functions based on the prerender functions. The difference is that it only updates the area that is in avatar's sight and not all the minimap. So far it looks to perform much better. The `createMapSurface()` seems to create the performance issues.

r-a-cristian-93 commented:
Hi @dorkster!

I have made some new changes:

- Tiles that are TILE_HIDDEN will not be rendered at all for all types of fog of way.
- When using TYPE_OVERLAY, TILE_HIDDEN_BOUNDARY is used to define be boundary tiles near TILE_HIDDEN and will be rendered on top.

This two above changes seem to have improved the performance. Still, there is much to be done as tall tiles are still sticking out through the fog of war. 
I think that the visited areas could be done similarly to have a boundary type tile.

![fow_foggy](https://user-images.githubusercontent.com/69315110/128413712-9585eea0-4823-4320-a6cb-8e82480a86fe.jpg)

r-a-cristian-93 commented:
Hi, @dorkster ! 

I made some research and i found a different algorithm for fog of war.

 [This](http://bobkoon.com/how-to-implement-a-fog-of-war-part-2-smooth/) is the original article that describes how it works . [And this ](https://github.com/r-a-cristian-93/flare-engine/tree/fow_bits) is my attempt at implementing it.
I handled the tall tiles that were sticking out by rendering the entire fog of war layer. All tiles that are not visited are not rendered except for those at the boundary in order to spare some processing power.

I think the tile masks should be defined by the modders in a text file and should not be hardcoded. This will give them the freedom to make the mask as detailed as they want and as big as they want.

If you believe that this approach is better and it's worth developing further please let me know.

![fow_bits](https://user-images.githubusercontent.com/69315110/130352898-ad2689f6-bc9d-4faf-a119-0817968ad810.jpg)

dorkster commented:
I like how the new method looks in motion (no tile "pop-in"). I'd say it's worth pursuing.

r-a-cristian-93 commented:
Hi, @dorkster!

[I have fixed some bugs for the fog of war. ](https://github.com/r-a-cristian-93/flare-engine/tree/fow_bits)
1. Some tiles that were not rendered (even if they should have been) now are rendered.
2. Tall tiles like trees and walls that were showing up like from nowhere how behave correctly.

Both issues were solved with the same fix. When deciding not to render some tile the game will check for it's corners. If any of them is visible that tile will be rendered.

On some maps, for example Stormrock Pass, some tiles are not rendered correctly. I can't figure out why this is happening. Maybe you could point me to the right direction.

![fow_stormrock_pass](https://user-images.githubusercontent.com/69315110/133801744-c39815ae-92ce-47c4-8258-24313a0f20f2.jpg)

dorkster commented:
Looking good. This should fix the missing layers: https://github.com/dorkster/flare-engine/commit/9e2ea71d1e41144178f28f94fcbf1ac210e2e52b

r-a-cristian-93 commented:
> Looking good. This should fix the missing layers: [dorkster@9e2ea71](https://github.com/dorkster/flare-engine/commit/9e2ea71d1e41144178f28f94fcbf1ac210e2e52b)

Thank you! That fixed it, and now i can see what was wrong.

dorkster commented:
This seems to fix it: https://github.com/dorkster/flare-engine/commit/848cb1bd0a5be1588637ac4983817f0a47ecc058

r-a-cristian-93 commented:
That indeed fixed the problem. 

> Known issues:
> - There is still some flicker in the transition between "red" and
>   "black" fow tiles
> - The game crashes without the fogofwar_ortho mod loaded

- The flicker disappeared when i updated the assets. 5345b799c30b3979015506eada9dbc151b1693e4
- The game doesn't crash any more in the absence of the mod. 7ff8531d1777077e2b89dd035c8c597f677160e9

So far everything seems to work fine. The only issue that remains is  the edge of the map where tall objects are present.


![poly_ok](https://user-images.githubusercontent.com/69315110/135710964-02741920-a3ae-4049-a823-d2bac201f33a.jpg)

![map_edge](https://user-images.githubusercontent.com/69315110/135711068-7aecd5c4-50a5-4e45-8bdb-ed765dbdba01.jpg)

r-a-cristian-93 commented:
Hi @dorkster!

I have updated the engine/fog_mask.txt file to better match other configuration files. I have also added [documentation ](https://github.com/r-a-cristian-93/flare-engine/wiki/Attribute-Reference#fogofwar)for the fog of war system.

I have some doubts about the **mask.data** definition as it doesn't really match the **raw** valuetype. Maybe a new type should be created for this attribute.
Also, in a map configuration a layer can be added to prededefine visited areas. This will also require a slightly different valuetype that **raw**.

dorkster commented:
I think `raw` is appropriate for the `mask.data` property. `raw` is sort of a catch-all for data that contains line breaks, which the `mask.data` property does. It also shouldn't be `raw(predefined_string)`. The parenthesis syntax is only used for `list()` and `repeatable()`.

One bug I noticed was with the paths for saved map fog data. If the map is located in anything other than `maps/`, saving/loading the data will fail. You can see this happen with flare-game's Halls of Infinity maps, which are stored in `maps/halls_of_infinity/`. I think the solution would be to hash (`Utils::hashString()`) the map paths when creating the saved fog data filename.

In my opinion, I don't think the mods should be part of this merge. In fact, I think we should create a separate repo for them.

By the way, I think this pull request is targeting the wrong branch. It should be [fow_bits](https://github.com/r-a-cristian-93/flare-engine/tree/fow_bits), right? Also, is it okay if I squash the commits down into one? I'll take care of all the issues I mentioned above, as well.

r-a-cristian-93 commented:
The mods were created only for reference and it's ok to move them to a new repository. As for the commits i would like to keep them separated and not squash them into one.

Now the branch should have been [fow_bits](https://github.com/r-a-cristian-93/flare-engine/tree/fow_bits). I continued the discussion here to just to keep it simple until the fog of war is complete. Let me know if i should create another pull request for the right branch.

dorkster commented:
>As for the commits i would like to keep them separated and not squash them into one.

Fair enough. It would be easier to remove the mod content from the history with everything squashed, but I should be able to do it without squashing.

No need to open another PR. I'll take care of the merge. I just wanted to make sure I was merging the correct branch.

dorkster commented:
@r-a-cristian-93 One more question: I've so far rebased everything on top of master and removed the mod content. Looking at the git log (see below), I see some sequences of commits that have the same title, such as **Separate fow dark layer(unvisited) and fog layer(visited)**. Were these intended to be singular commits? If so, I think they should be combined.

```
14874faa Changed fow_mask.txt format to better match other configuration files
631e717e Moved documentation to correct location.
a63a55e9 Let the modder define the colors of fog of war type 2-tint tiles.
2de158b8 Let the modder name the fow tileset file.
ca377e2e Added fow documentation.
4e53d377 Fix game hanging when mod is not enabled
30b84d31 FogOfWar: Attempt to fix corruption when rendering ortho maps
d1121d02 Fixed some bugs, created others
e1d799e3 Allow map files to override engine setting 'fogofwar' and 'save_fogofwar'
4ab3d7b2 Fix missing iso layers when using FogOfWar::TYPE_OVERLAY
a1f3a0d3 Fix Segmentation Fault for some tiles
7b01bcd4 FOW. Fixed rendering of missing tiles. Trees and other tall objects do not show up suddenly anymore
b0b93561 Skip rendering fog tiles that are underneath dark tiles
20fa6ca2 Removed commented out code
2479ea7f Removed fow tile_numbers sprites. TILE_HIDDEN defined as static member and proper fow initialization.
85184835 Removed Avatar's sight property
0dd1f3ba Load fow mask from file
80f7c819 Load fow mask from file
f94881f9 Load fow mask from file
3d4962cd Improved skipped tiles from rendering
4c400eb9 Separate fow dark layer(unvisited) and fog layer(visited)
574c29bb Separate fow dark layer(unvisited) and fog layer(visited)
b138b412 Separate fow dark layer(unvisited) and fog layer(visited)
61d3dfbd Separate fow dark layer(unvisited) and fog layer(visited)
895640a4 Separate fow dark layer(unvisited) and fog layer(visited)
b4bff9e6 Separate fow dark layer(unvisited) and fog layer(visited)
843249ce Fixed missplaced tiles
151f2b46 Debug tile numbers
97d7e3ee Bitwise operations tiles
551a67f6 Fog of war hidden state boundary
45e63086 Fog of war hidden state boundary
807d1fcb Named the tile state
4c46f244 Named the tile state
6bc39f49 No need to verify delta
d2c1873a skip rendering tiles if never visited
b9001150 Improved minimap performance for fog of war
5f17d38b optimize minimap update for fog of war
2c676c31 Undone unuseful change
2ee5c4ed Undone unuseful change
5f324b89 Undone unuseful change
cb5ca4f4 Added fog of war TYPE_MINIMAP
1978c771 Merged dorkster's code - fow by tinting tiles
fdfba9f4 Merged dorkster's code - fow by tinting tiles
f1dfd8a1 Prerender minimap only when the avatar moves or teleports
47021b0e Prerender minimap only when the avatar moves
28546621 Fog of war for minimap
f388ac0c Better code for hiding out of sight entities
6f638e48 Hide loot and its tooltip if out of sight
dd89a952 Skip rendering out of sight entities
7b941d5c FogOfWar handles intramap teleport
470b8dc8 Save fog of war on Save button and on map change
62105ce9 Save fog of war layer data implemented
2bff0ce4 Moved fow resources to FogOfWar class. Added engine setting fogofwar and save_fogofwar
8c00095c Renamed FogOfWar function update() to logic();
9ac798fc FogOfWar class refactoring
48f7cc78 Revert to FogOfWarClass
54c1c267 Removed duplicate code and changed renderIsoLayer() function
506bd25f Update fogofwar when changhing map
ea3e4b35 Fogofwar tile type: 0=in_sight 1=unexplored 2=out_of_sight
6a30c165 no need to check the fogofwar tile state
1d413e97 tileset_fogofwar initialisation  moved to a more suitable place
4e164ee5 Moved fogofwar code to a function. Limited iterations only to surrounding area.
d3dc7df7 Different fogofwar Map_Layer for tiles and for explored/unexplored
4e4fef03 FogOfWar
140d72ee FogOfWar
9c08efcd FogOfWar class removed. fogofwar implementation as map layer
9d08f95f Added FogOfWar class
```

r-a-cristian-93 commented:
Mostly the commits that have the same title are to fix some small mistakes i did with the previous commit. So, if you think it's better to combine them then you should. I personally don't know how to do it, but i would like to.

dorkster commented:
>I personally don't know how to do it, but i would like to.

Typically, I will use `git rebase -i` to perform an interactive rebase. In this case, I want to rebase on top of master, so the command is `git rebase -i master`. This creates a list of all the commits on top of master in your $EDITOR, each marked as `pick`.  To combine a commit with the previous one, you can change `pick` to `squash` or `fixup`. There are lots of other actions you can do, which you can see commented at the bottom of the commit list.

dorkster commented:
And... merged! The mods repo can be found here: https://github.com/flareteam/flare-mod-fogofwar

WithinAmesia commented:
How is this set up using: https://github.com/flareteam/flare-mod-fogofwar and how can the radius become 10 up from 8? Do I have to make maps differently in Tiled for this to work with Flare. If there are needed details around map creation / running this I would much appreciate a simple list / tips on how to set this up. It looks great :-).

WithinAmesia commented:
Okay I am kind of lost on how to get this to work. I tried to get a fresh 1.12 Flare working with this mod and I did not get it work. Here is what I did:

1.) I downloaded this: https://github.com/flareteam/flare-mod-fogofwar and unpacked it.
2.) I made a fresh 1.12 Flare install and put the "fogofwar" from the upacked.zip (from flare-mod-fogofwar-master\mods\fogowar) into the mods file in the Fresh 1.12 Flare
3.) I ran the fresh 1.12 Flare and went to configure and enabled the fogowar mod on and made a new fresh character
4.) I played a the fresh character through the into of the empyrean campaign and there was no fog of war.

What am I missing and what does this fog of war mod require to operate?
I am hoping to use this to make a dungeon with and I'd like to get this working with 11 tile radius and if this requires extra tiled / Flare map and or engine tweaking I would like to understand how to get this to work before making a bunch of areas and potentially having to go through retrofit nightmares XD.

r-a-cristian-93 commented:
@WithinAmesia Fog of war has been added in version 1.13 which is not released yet. You would have to build it yourself for now. Maybe i could send you a build with the latest features if you would tell me what operating system you have.

WithinAmnesia commented:
@r-a-cristian-93 I have Windows 10 Home. I am trying to make a ~'super dungeon' prototype while 1.13 is being worked on and the Fog of War I hope to include in the prototype (thus I am asking around here XD). I just got the new / super dungeon tileset to work yesterday: https://www.deviantart.com/withinamnesia/art/Super-Dungeon-First-898068459 + https://opengameart.org/content/flare-super-dungeon-tileset-version-10 . Although before I build out ~33 x 1000x1000 tile levels I would like to have fog of war working to avoid any potential retrofit nightmares (retrofitting massive areas is hard, so even the possibility is scary x.X) and I think the fog of war really improves the gameplay experience :-D.

r-a-cristian-93 commented:
@WithinAmesia Here you can find the windows binary with the latest features https://github.com/r-a-cristian-93/flare-engine/releases/download/v1.12/flare-win32-v1.12.16

WithinAmnesia commented:
@r-a-cristian-93 I think the file is 981 KB (1,004,308 bytes) and with no extension. Was this supposed to be an installer .exe like the older version of 1.12 like "flare-win32-v1.12.exe" with ~132,188 KB (135,359,860) bytes? What am I missing?

r-a-cristian-93 commented:
@WithinAmesia it was a zip file but somehow the extension was missing. I fixed it, try to download again.

WithinAmnesia commented:
@r-a-cristian-93 The file crashes on start up saying it is missing some .dll or something. It looks like the flare.exe after a fun install like if I installed flare then I just copied only the flare.exe and not the whole folder that makes it all work?

WithinAmnesia commented:
@r-a-cristian-93 I got it to work. I took a fresh install of 1.12 official release and swapped out the flare.exe for the one you sent in the zip and the fog of war works.

WithinAmnesia commented:
https://www.deviantart.com/withinamnesia/art/9-FPS-Hmm-898339783 Bad news my game now lags and chops frames with as low as 9 FPS in a 1000x1000 tile level with Fog of War. Yet I did not need any extra work to have it work. So I can get the dungeon level work started :-D. Although I hope I can have Fog of War work in the future. It is somehow currently reducing the frame rate by ~7 times when the player character is running into the Fog of War. It is cool but the game starts lagging and skipping frames on a beastly rig so god help a regular PC XD.

r-a-cristian-93 commented:
You should set the renderer to sdl_hardware. If you already did that the problem could be your high screen resolution. The more tiles are visible the slower the game will be. Also, play around with flare's video settings. Or could be something else. If you share your mod on github would be easier to help you

dorkster commented:
@WithinAmesia Even on 1.12, I get about 13 FPS when playing a 1000x1000 test map. Turning off the minimap gets me back up to 60. I also tested with the latest master and the fogofwar mod. Again, the turning the minimap off helps, but there were also drops when uncovering completely hidden areas of the map (no issue with tiles that were already discovered).

So I think the minimap with or wthout fog of war is the biggest performance hog, and as such should get optimization attention first.

WithinAmnesia commented:
https://www.deviantart.com/withinamnesia/art/7-FPS-Yikes-898408293?ga_submit_new=10%3A1637369145 7 FPS Yikes (with sdl_hardware; it was always one before too). This is with the mini map turned off too. I tried x2 zoom mini map and that lags the most as well.

I have an idea. Can the Fog of War have (a potential version?) just the dark and light radius? Like no black areas just like everything is grey like it is already explored and the player still has a light radius? The Fog of War works great with the explored areas it just slows way down with uncovering the black unexplored areas. I like the light radius mechanics and it avoid screen sniping corners to reduce the cheese strats XD.

I will make a .zip of my messing test build and upload it, somewhere? I will make a Mega.nz I suppose and give a link here hopefully for the time being?

WithinAmnesia commented:
Also off topic kinda. Is the feature where the enemies have the same attribute reference / exec commands as events in the 1.12.16.exe? I am wondering this in case I can tinker around with spawning enemies on maps base upon character level range to also work on experience scaling enemies based upon level range (100% 50% 10% experience versions of the same enemy but the higher the level at a certain range the monster spawns into 50% and then 10% experience to simulate experience scaling).

WithinAmnesia commented:
https://mega.nz/file/SBpCQYjZ#cy9FtGYJAgad1K9xY6NTqLgGbdZ6zQJMS3epjShq-8Y The current testbed I am using. The 1000x1000 'Super Dungeon' in the screenshots area is named "Super_Dungeon_Test.txt".

dorkster commented:
Not to stomp on your creativity, but I feel it's worth mentioning that 1000x1000 maps aren't something we designed for. Even Tiled choked a bit for me when trying to edit it with automapping. There's a reason games divide their worlds up into levels (or in the case of open-world games, "cells"). There's obviously performance gains to be made here. But even with said gains, smaller maps will perform better.

>Also off topic kinda. Is the feature where the enemies have the same attribute reference / exec commands as events in the 1.12.16.exe?

Yes it is, so you should be able to use `requires_level` and `requires_not_level` on enemy groups.

WithinAmnesia commented:
I found this: https://www.reddit.com/r/Minecraft/comments/gwd84/new_circle_guide_131_even_diameters_bigger_no/ So I suppose this could be used to adjust as a template the light radius diameter.

@dorkster Yes the Flare engine runs way better on smaller maps but I want to test a 1000x1000 tile map with all of the features and see if my best computer can run it and if so over the years by Moore's Law my beastly rig will become that of a refrigerator OS in equivalency. I am expecting this game to take a few years to make and I hope that it can be a special game and last the test of time like Diablo II, Baldur's Gate, WoW vanilla but I do run a risk here. I have not tested a fully fleshed out 1000x1000 map with enemies and other performance draws (such as animated tiles potentially).

I have a few computers to test the 1000x1000 map on ranging from potato to aircraft carrier so I will have to take a gamble and see if 1000x1000 will melt the average PC once all the enemies and stress test can be fleshed out. If the average PC works within tolerance of frame rate (not choppy) I can be happy with that. Although if it does not work than well I cannot have areas that are literally unplayable lol. I am hoping in some fashion to use this game as a template for a larger game as a distant goal and I am trying to find a way to translate my life experience of working in massive forests and hiking up difficult mountains into the gameplay so I am hoping that the 1000x1000 areas can in a small way replicate that feeling but hopefully still not melt PCs XD.
